**Importing Python Libraries**

In [2]:
import pandas as pd
import numpy as np

**Loading the foodBalanceSheet_E_africa Data**

In [3]:
from google.colab import files
uploaded = files.upload()

Saving FoodBalanceSheets_E_Africa_NOFLAG.csv to FoodBalanceSheets_E_Africa_NOFLAG.csv


In [4]:
#To store dataset in pandas DataFrame
food = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', encoding = "latin-1")
food.head()


Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [5]:
food.shape

(60943, 12)

Question 1
**What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?** 

In [6]:
food.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

The mean across the whole dataset in 2017 = 140.92
The S.D across the whole dataset in 2017 = 1671.86

**Question 2; Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?**

In [7]:
#Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?
Area_sum = food.groupby('Area').sum()
largest_sum = Area_sum['Y2017'].idxmax()
print("Area with the largest sum is:", largest_sum)

Area with the largest sum is: Nigeria


<ipython-input-7-64f8fe52707e>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  Area_sum = food.groupby('Area').sum()


**Question 3; A pandas Dataframe with dimensions (100,3) has how many features and observations?**

In [8]:
#Answer = 100 features and 3 observations

**Question 4; What is the total sum of Wine produced in 2015 and 2018 respectively?**

In [9]:
Wine_produced = food.groupby('Item').sum().loc['Wine'][['Y2015','Y2018']]
print(Wine_produced)

Y2015    4251.81
Y2018    4039.32
Name: Wine, dtype: float64


<ipython-input-9-04c48b0ce969>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  Wine_produced = food.groupby('Item').sum().loc['Wine'][['Y2015','Y2018']]


**Question 5; Given the following numpy array 
array  = ([[94, 89, 63],
             [93, 92, 48],
             [92, 94, 56]])
How would you select  the elements in bold and italics from the array?**

In [10]:
import numpy as np
array = np.array([[94, 89, 63], [93, 92, 48], [92, 94, 56]])
array[ : 2, 1 : ] 

array([[89, 63],
       [92, 48]])

**Question 6; Which of these python data structures is unorderly?**

In [11]:
#Set is an unordered collection of unique items. Hence, duplicate or repitive elements are not allowed in sets as they represent a single entity. 

**Question 7; What is the total number of unique countries in the dataset**

In [12]:
print("The total numer of unique countries is:", food['Area'].nunique())


The total numer of unique countries is: 49


**Question 8; Consider the following list of tuples:
y = [(2, 4), (7, 8), (1, 5, 9)]
How would you assign element 8 from the list to a variable x?**

In [13]:
y = [(2, 4), (7, 8), (1, 5, 9)]
x = y[1][-1]
x

8

**Question 9;Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?**

In [14]:
years= ['Y2014', 'Y2015', 'Y2016','Y2017', 'Y2018']
food.groupby('Element').sum().loc[['Stock Variation'], years].idxmax(axis=1)

<ipython-input-14-13dc8a91bb93>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  food.groupby('Element').sum().loc[['Stock Variation'], years].idxmax(axis=1)


Element
Stock Variation    Y2014
dtype: object

**Question 10; Which of the following is a python inbuilt module?**

In [ ]:
#Math is a math module python used in extending mathematical functions

**Question 11; Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?**

In [15]:
year = 'Y2017'
Element = food.groupby('Element').sum().loc['Processing'], (year)
Element

<ipython-input-15-d9d9f0b134ff>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  Element = food.groupby('Element').sum().loc['Processing'], (year)


(Area Code         271940.0
 Item Code        5350416.0
 Element Code    10313310.0
 Y2014             282923.0
 Y2015             287929.0
 Y2016             280631.0
 Y2017             292836.0
 Y2018             308429.0
 Name: Processing, dtype: float64,
 'Y2017')

**Question 12;What is the total number and percentage of missing data in 2014 to 3 decimal places?**

In [16]:
print("The sum of missing data in 2014 is :", food['Y2014'].isna().sum())


The sum of missing data in 2014 is : 1589


In [17]:
#The percentage of the missing data is the division of the sum of the missing data by the total number of rows
print("The percentage of missing data in 2014 is:", (round(food['Y2014'].isna().sum() / 60943 * 100, 3)))

The percentage of missing data in 2014 is: 2.607


**Question 13;Which of the following dataframe methods can be used to access elements across rows and columns**

In [18]:
#df.iloc( )

**Question 14;Which year had the least correlation with ‘Element Code’?**

In [19]:
years= ['Y2014', 'Y2015', 'Y2016','Y2017', 'Y2018']
cor=food.corr().abs()    
print(cor['Element Code'][years].idxmin())

Y2016


<ipython-input-19-a59a2867eca4>:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  cor=food.corr().abs()


**Question 15;If you have the following list
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]
col = [‘Age’,’Nationality’,’Overall’]
How do you create a pandas DataFrame using this list, to look like the table below?**

In [20]:
#None of the above

**Question 16,Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?**

In [21]:
lowest = food.groupby('Area')['Y2017'].sum()
lowest_sum = lowest.idxmin()


def lowest_number(lowest_sum):
  the_minimum = index.min(lowest_sum)
  lowest.sort()
  print(lowest_sum[6])
lowest_sum

'Ethiopia PDR'

**Question 17, What would be the output for?
S = [['him', 'sell'], [90, 28, 43]]
S[0][1][1]**

In [22]:
S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

'e'

**Question 18, How would you check for the number of rows and columns in a pandas DataFrame named df?**

In [23]:
food.shape
#df.shape

(60943, 12)

**Question 19, Given the following python code, what would the output of the code give?
my_tuppy = (1,2,5,8)
my_tuppy[2] = 6**

In [24]:
my_tuppy = (1,2,5,8) my_tuppy[2] = 6

SyntaxError: ignored

**Question 20, What is the total Protein supply quantity in Madagascar in 2015**

In [34]:
#protein = food[food['Element']=='Protein supply quantity (g/capita/day)'].groupby(["Area", "Y2015"])
#protein_supply = food.groupby('Element').sum().loc['Protein supply quantity']['Y2015']
#protein = food[food['Element']=='Protein supply quantity']
year = 'Y2015'
area = 'Madagascar'
Element = food.groupby('Element').sum().loc['Protein supply quantity (g/capita/day)'], (year), (area)
Element



<ipython-input-34-a9f08735ca34>:6: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  Element = food.groupby('Element').sum().loc['Protein supply quantity (g/capita/day)'], (year), (area)


(Area Code         675050.00
 Item Code       13535000.00
 Element Code     3385502.00
 Y2014              11836.46
 Y2015              11833.95
 Y2016              11779.69
 Y2017              11842.45
 Y2018              11833.56
 Name: Protein supply quantity (g/capita/day), dtype: float64,
 'Y2015',
 'Madagascar')